crates/agent/src/lib.rs

```diff
@@ -32,33 +55,107 @@ pub async fn serve<E>(
+    let handlers_by_table = handlers
+        .into_iter()
+        .map(|h| (h.table_name(), Arc::new(Mutex::new(h))))
+        .collect::<HashMap<_, _>>();
+
+    let mut listener = PgListener::connect_with(&pg_pool).await?;
+
+    listener.listen(AGENT_NOTIFICATION_CHANNEL).await?;
+
+    // We use a channel here for two reasons:
+    // 1. Because handlers run one task at a time, and can also indicate that they have more work to perform or not,
+    //    we want to balance the time spent processing each type of handler so that no one handler can monopolize resources.
+    // 2. It makes it easy to preemptively schedule at least one run of each handler on boot up to allow for handling requests
+    //    that came in while we weren't running
+    // NOTE: it is critical that we use an unbounded channel here, otherwise we would open ourselves up to a deadlock scenario
+    // Example deadlock when using a bounded channel:
+    // 1. A spike of work comes in
+    let (task_tx, mut task_rx) = mpsc::unbounded_channel::<String>();
+
+    // Each handler gets run at least once to check if there is any pending work
+    for (handler_table, _) in handlers_by_table.iter() {
+        task_tx.send(handler_table.to_string())?;
+    }
+
+    let task_tx_cloned = task_tx.clone();
+    let listen_to_datbase_notifications = async move {
+        loop {
+            let item = listener.recv().await.map_err(|e| anyhow::Error::from(e))?;
+            let notification: AgentNotification = serde_json::from_str(item.payload())
+                .context("deserializing agent task notification")?;
+            let table: &str = &notification.table;
+
+            tracing::debug!(table = table, "Message received to invoke handler");
+            task_tx_cloned
+                .send(table.to_string())
+                .map_err(|e| anyhow::Error::from(e))?
+        }
+        // Need this here to indicate that this future returns an anyhow::Result<()>
+        #[allow(unreachable_code)]
+        Ok::<(), anyhow::Error>(())
+    };
+
+    let mut listener_handler = tokio::spawn(listen_to_datbase_notifications);
 
     tokio::pin!(exit);
+
     loop {
-        // Pick handler with the next deadline.
-        let (handler, deadline) = handlers
-            .iter_mut()
-            .min_by_key(|i| i.1)
-            .expect("handlers is not empty");
-
-        // Sleep until its deadline has elapsed.
-        let sleep = tokio::time::sleep_until(*deadline);
         tokio::select! {
             _ = &mut exit => {
                 tracing::debug!("caught signal; exiting...");
                 return Ok(()) // All done.
             }
-            _ = sleep => (),
-        };
+            listener_res = &mut listener_handler => {
+                match listener_res {
+                    // It should be impossible to get here since `listen_to_datbase_notifications` loop has no `return`s or `break`s
```

Comment:
nit: use `unreachable!("the reason")` then? External conditions which can fail should clearly be errors, but it's better IMO to panic immediately if an implementation invariant is violated because it gives a clearer stack trace. Returning an error tends to muddle exactly where the error originated from.